### ISSUE TITLE
Global initialization is flaky

### ISSUE THREAD

**Issue by steven-murray** [21cmfast/21cmFAST#114]
Continuing discussion from #113:

>  There are some of these high-level global initialisation routines which we have to think more carefully about. My fix here was to run init_ps for every time we initialise photon cons, but that re-runs things that sometimes don't need to be. We've also run in to this error a few times when code's added, because it's very easy to forget that a function is dependent on init_ps having already been run, and assume that it has been.

Do we need to make this a more robust procedure?

**Comment by BradGreig** [comment]
@steven-murray, adding it as you did to the initialisation of the photon conservation should be sufficient. It'll (at worse) result in one additional call of init_ps() which is negligible in terms of total computation time.

My only alternative suggestion to this would be to write a small standalone python callable C function, which performs init_ps. Then, in the python wrappers for `run_coeval` or `run_lightcone`, when it checks to do photon cons, it can check if its reading from cache and perform the init_ps calculation by calling this function. I'm not sure if this is any more robust, but by separating it out it may make it easier to fix and deal with in future.

As an aside, note that currently the `PHOTON_CONS` flag is only useable for `run_coeval` and `run_lightcone`. If the user attempted to call `ionized_box` it'll fail. Perhaps `ionized_box` should also have the ability to internally setup the photon non-conservation correction? In doing so, I suspect that might be a little more complicated.

**Comment by steven-murray** [comment]
Could we do something like have a `power_spectrum` global struct instance, which gets inited with `init_ps`, and also has a bool field called `initialized`? Then from any function which requires it to be initialized, we could just do `if(~power_spectrum.initialized) init_ps()`. We could _also_ make the `init_ps` function available at python-level (this should be trivial), but if we ever want to make this work directly from C again, we probably want to make sure the individual functions are consistent, rather than having to defer to python.

On your note about PHOTON_CONS -- I think I'll make a new issue for that, to be tackled in v3.1 or so.

**Comment by BradGreig** [comment]
Yeah, something like that would work, though I'd consider renaming the instance to something else (to avoid naming confusion).

**Comment by steven-murray** [comment]
This might be something that has to be done at the top of high-level function like `run_coeval` (and freed at the end).
